### ISSUE TITLE
Concretizer fails with unhelpful error message when a dag is given as dependency

### ISSUE THREAD

**Issue by s-sajid-ali** [spack/spack#13834]
Trying to concretize `fftw+mpi ^/{mpich-dag}` fails with unhelpful error. 

### Steps to reproduce the issue

```
[sas4990@quser10 ~]$ spack find -lv mpich
==> 1 installed package
-- linux-rhel7-ivybridge / gcc@8.3.0 ----------------------------
btlyk64 mpich@3.3.2 device=ch4 +hydra netmod=ucx +pci pmi=pmi2 +romio+slurm~verbs+wrapperrpath
[sas4990@quser10 ~]$ spack spec -I fftw+mpi ^/btlyk64 %gcc@8.3.0
Input spec
--------------------------------
 -   fftw+mpi
[+]      ^mpich@3.3.2%gcc@8.3.0 device=ch4 +hydra netmod=ucx +pci pmi=pmi2 +romio+slurm~verbs+wrapperrpath arch=linux-rhel7-ivybridge
[+]          ^libpciaccess@0.13.5%gcc@8.3.0 arch=linux-rhel7-ivybridge
[+]          ^libxml2@2.9.9%gcc@8.3.0~python arch=linux-rhel7-ivybridge
[+]              ^libiconv@1.16%gcc@8.3.0 arch=linux-rhel7-ivybridge
[+]              ^xz@5.2.4%gcc@8.3.0 arch=linux-rhel7-ivybridge
[+]              ^zlib@1.2.11%gcc@8.3.0+optimize+pic+shared arch=linux-rhel7-ivybridge
[+]          ^slurm@19-05-3-2%gcc@8.3.0~gtk~hdf5~hwloc~mariadb~pmix+readline arch=linux-rhel7-ivybridge

Concretized
--------------------------------
 -   fftw@3.3.8%gcc@8.3.0+mpi~openmp~pfft_patches precision=double,float arch=linux-rhel7-ivybridge
[+]      ^mpich@3.3.2%gcc@8.3.0 device=ch4 +hydra netmod=ucx +pci pmi=pmi2 +romio+slurm~verbs+wrapperrpath arch=linux-rhel7-ivybridge
[+]          ^libpciaccess@0.13.5%gcc@8.3.0 arch=linux-rhel7-ivybridge
[+]          ^libxml2@2.9.9%gcc@8.3.0~python arch=linux-rhel7-ivybridge
[+]              ^libiconv@1.16%gcc@8.3.0 arch=linux-rhel7-ivybridge
[+]              ^xz@5.2.4%gcc@8.3.0 arch=linux-rhel7-ivybridge
[+]              ^zlib@1.2.11%gcc@8.3.0+optimize+pic+shared arch=linux-rhel7-ivybridge
[+]          ^slurm@19-05-3-2%gcc@8.3.0~gtk~hdf5~hwloc~mariadb~pmix+readline arch=linux-rhel7-ivybridge

Input spec
--------------------------------
==> Error: 'NoneType' object has no attribute 'endswith'
[sas4990@quser10 ~]$
...
```

### Error Message
```
....
 -   fftw@3.3.8%gcc@8.3.0+mpi~openmp~pfft_patches precision=double,float arch=linux-rhel7-ivybridge
[+]      ^mpich@3.3.2%gcc@8.3.0 device=ch4 +hydra netmod=ucx +pci pmi=pmi2 +romio+slurm~verbs+wrapperrpath arch=linux-rhel7-ivybridge
[+]          ^libpciaccess@0.13.5%gcc@8.3.0 arch=linux-rhel7-ivybridge
[+]          ^libxml2@2.9.9%gcc@8.3.0~python arch=linux-rhel7-ivybridge
[+]              ^libiconv@1.16%gcc@8.3.0 arch=linux-rhel7-ivybridge
[+]              ^xz@5.2.4%gcc@8.3.0 arch=linux-rhel7-ivybridge
[+]              ^zlib@1.2.11%gcc@8.3.0+optimize+pic+shared arch=linux-rhel7-ivybridge
[+]          ^slurm@19-05-3-2%gcc@8.3.0~gtk~hdf5~hwloc~mariadb~pmix+readline arch=linux-rhel7-ivybridge

Input spec
--------------------------------
Traceback (most recent call last):
  File "/home/sas4990/packages/spack/bin/spack", line 64, in <module>
    sys.exit(spack.main.main())
  File "/home/sas4990/packages/spack/lib/spack/spack/main.py", line 715, in main
    return _invoke_command(command, parser, args, unknown)
  File "/home/sas4990/packages/spack/lib/spack/spack/main.py", line 457, in _invoke_command
    return_val = command(parser, args)
  File "/home/sas4990/packages/spack/lib/spack/spack/cmd/spec.py", line 79, in spec
    print(spec.tree(**kwargs))
  File "/home/sas4990/packages/spack/lib/spack/spack/spec.py", line 3836, in tree
    if node.package.installed_upstream:
  File "/home/sas4990/packages/spack/lib/spack/spack/spec.py", line 1157, in package
    self._package = spack.repo.get(self)
  File "/home/sas4990/packages/spack/lib/spack/spack/repo.py", line 1203, in get
    return path.get(spec)
  File "/home/sas4990/packages/spack/lib/spack/spack/repo.py", line 94, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/home/sas4990/packages/spack/lib/spack/spack/repo.py", line 653, in get
    return self.repo_for_pkg(spec).get(spec)
  File "/home/sas4990/packages/spack/lib/spack/spack/repo.py", line 94, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/home/sas4990/packages/spack/lib/spack/spack/repo.py", line 885, in get
    raise UnknownPackageError(spec.name)
  File "/home/sas4990/packages/spack/lib/spack/spack/repo.py", line 1293, in __init__
    if name.endswith(".yaml"):
AttributeError: 'NoneType' object has no attribute 'endswith'
...
```

However, everything works if I specify all the variants to the concretizer : 
```
[sas4990@quser10 ~]$ spack spec -I fftw ^mpich@3.3.2 device=ch4 netmod=ucx pmi=pmi2 +slurm  %gcc@8.3.0
Input spec
--------------------------------
 -   fftw
 -       ^mpich@3.3.2%gcc@8.3.0 device=ch4 netmod=ucx pmi=pmi2 +slurm

Concretized
--------------------------------
 -   fftw@3.3.8%gcc@8.3.0+mpi~openmp~pfft_patches precision=double,float arch=linux-rhel7-ivybridge
[+]      ^mpich@3.3.2%gcc@8.3.0 device=ch4 +hydra netmod=ucx +pci pmi=pmi2 +romio+slurm~verbs+wrapperrpath arch=linux-rhel7-ivybridge
[+]          ^findutils@4.6.0%gcc@8.3.0 patches=84b916c0bf8c51b7e7b28417692f0ad3e7030d1f3c248ba77c42ede5c1c5d11e,bd9e4e5cc280f9753ae14956c4e4aa17fe7a210f55dd6c84aa60b12d106d47a2 arch=linux-rhel7-ivybridge
[+]              ^autoconf@2.69%gcc@8.3.0 arch=linux-rhel7-ivybridge
[+]                  ^m4@1.4.18%gcc@8.3.0 patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00,fc9b61654a3ba1a8d6cd78ce087e7c96366c290bc8d2c299f09828d793b853c8 +sigsegv arch=linux-rhel7-ivybridge
[+]                      ^libsigsegv@2.12%gcc@8.3.0 arch=linux-rhel7-ivybridge
[+]                  ^perl@5.30.0%gcc@8.3.0+cpanm+shared+threads arch=linux-rhel7-ivybridge
[+]                      ^gdbm@1.18.1%gcc@8.3.0 arch=linux-rhel7-ivybridge
[+]                          ^readline@8.0%gcc@8.3.0 arch=linux-rhel7-ivybridge
[+]                              ^ncurses@6.1%gcc@8.3.0~symlinks~termlib arch=linux-rhel7-ivybridge
[+]                                  ^pkgconf@1.6.3%gcc@8.3.0 arch=linux-rhel7-ivybridge
[+]              ^automake@1.16.1%gcc@8.3.0 arch=linux-rhel7-ivybridge
[+]              ^libtool@2.4.6%gcc@8.3.0 arch=linux-rhel7-ivybridge
 -               ^texinfo@6.5%gcc@8.3.0 patches=12f6edb0c6b270b8c8dba2ce17998c580db01182d871ee32b7b6e4129bd1d23a,1732115f651cff98989cb0215d8f64da5e0f7911ebf0c13b064920f088f2ffe1 arch=linux-rhel7-ivybridge
[+]          ^libpciaccess@0.13.5%gcc@8.3.0 arch=linux-rhel7-ivybridge
[+]              ^util-macros@1.19.1%gcc@8.3.0 arch=linux-rhel7-ivybridge
[+]          ^libxml2@2.9.9%gcc@8.3.0~python arch=linux-rhel7-ivybridge
[+]              ^libiconv@1.16%gcc@8.3.0 arch=linux-rhel7-ivybridge
[+]              ^xz@5.2.4%gcc@8.3.0 arch=linux-rhel7-ivybridge
[+]              ^zlib@1.2.11%gcc@8.3.0+optimize+pic+shared arch=linux-rhel7-ivybridge
[+]          ^slurm@19-05-3-2%gcc@8.3.0~gtk~hdf5~hwloc~mariadb~pmix+readline arch=linux-rhel7-ivybridge

[sas4990@quser10 ~]$
```

### Information on your system

 1. Platform :  `linux-rhel7-ivybridge`
 2. Custom slurm in `packages.yaml`

If this is a known issue, feel free to close it.

**Comment by alalazo** [comment]
I think the error is on your side. Also on the latest `develop` the error message seems to be improved with respect to what you report. Which version or commit hash of Spack are you using?

The problem should be solved if you remove the trailing `%gcc@8.3.0`. Since a spec specified by its hash is concrete it doesn't lack any information, including compiler. The trailing `%gcc@8.3.0` is then interpreted by the parser as an anonimous spec of its own. Here's for reference what I get on my machine at 9a6a19d464f31d301fb333c5110042f9ead5d501:
```console
$ spack find -l openmpi
==> 1 installed package
-- linux-ubuntu18.04-broadwell / gcc@9.0.1 ----------------------
gzhhrhc openmpi@3.1.4

$ spack spec fftw ^/gzhhrhc
[ ... ]
Concretized
--------------------------------
fftw@3.3.8%gcc@9.0.1+mpi~openmp~pfft_patches precision=double,float arch=linux-ubuntu18.04-broadwell
    ^openmpi@3.1.4%gcc@9.0.1~cuda+cxx_exceptions fabrics=none ~java~legacylaunchers~memchecker~pmi schedulers=none ~sqlite3~thread_multiple+vt arch=linux-ubuntu18.04-broadwell
        ^hwloc@1.11.11%gcc@9.0.1~cairo~cuda~gl+libxml2~nvml+pci+shared arch=linux-ubuntu18.04-broadwell
            ^libpciaccess@0.13.5%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
            ^libxml2@2.9.9%gcc@9.0.1~python arch=linux-ubuntu18.04-broadwell
                ^libiconv@1.16%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
                ^xz@5.2.4%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
                ^zlib@1.2.11%gcc@9.0.1+optimize+pic+shared arch=linux-ubuntu18.04-broadwell
            ^numactl@2.0.12%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell

$ spack spec fftw ^/gzhhrhc %gcc@9.0.1
[ ... ]
Input spec
--------------------------------
%gcc@9.0.1

Concretized
--------------------------------
==> Error: Attempting to concretize anonymous spec
```

**Comment by s-sajid-ali** [comment]
Thanks for the tip!
